ktlint-ruleset-standard/src/test/kotlin/com/pinterest/ktlint/ruleset/standard/rules/ChainMethodContinuationRuleTest.kt

```diff
@@ -0,0 +1,811 @@
+        chainMethodContinuationRuleAssertThat(code)
+            .addAdditionalRuleProvider { IndentationRule() }
+            .hasLintViolations(
+                LintViolation(1, 27, "Expected newline before '.'"),
+                LintViolation(4, 5, "Unexpected newline before '.'"),
+                LintViolation(6, 27, "Expected newline before '.'"),
+                LintViolation(9, 1, "Unexpected newline before '.'"),
+            ).isFormattedAs(formattedCode)
+    }
+
+    @Test
+    fun `Given a multiline method chain which is correctly wrapped`() {
+        val code =
+            """
+            val foo1 = listOf(1, 2, 3)
+                .filter { it > 2 }
+                .sum()
+            val foo2 = listOf(1, 2, 3)
+                .filter { it > 2 }!!
+                .sum()
+            val foo3 = listOf(1, 2, 3)
+            .filter { it > 2 }!!
+                ?.sum()
+            """.trimIndent()
+        chainMethodContinuationRuleAssertThat(code).hasNoLintViolations()
+    }
+
+    @Test
+    fun `Given a multiline method chain which is not correctly wrapped`() {
+        val code =
+            """
+            val foo1 = listOf(1, 2, 3).filter { it > 2 }
+                .sum()
+            val foo2 = listOf(1, 2, 3).filter { it > 2 }!!
+                .sum()
+            val foo3 = listOf(1, 2, 3).filter { it > 2 }!!
+                ?.sum()
+            """.trimIndent()
+        val formattedCode =
+            """
+            val foo1 = listOf(1, 2, 3)
+                .filter { it > 2 }
+                .sum()
+            val foo2 = listOf(1, 2, 3)
+                .filter { it > 2 }!!
+                .sum()
+            val foo3 = listOf(1, 2, 3)
+                .filter { it > 2 }!!
+                ?.sum()
+            """.trimIndent()
+        chainMethodContinuationRuleAssertThat(code)
+            .hasLintViolations(
+                LintViolation(1, 27, "Expected newline before '.'"),
+                LintViolation(3, 27, "Expected newline before '.'"),
+                LintViolation(5, 27, "Expected newline before '.'"),
+            ).isFormattedAs(formattedCode)
+    }
+
+    @Test
+    fun `Given next chain which is incorrectly wrapped chain operator in object declaration`() {
+        val code =
+            """
+            val foo = object : Runnable {
+                override fun run() {
+                    /* no-op */
+                }
+            }.
+                run()
+            """.trimIndent()
+        val formattedCode =
+            """
+            val foo = object : Runnable {
+                override fun run() {
+                    /* no-op */
+                }
+            }.run()
+            """.trimIndent()
+        chainMethodContinuationRuleAssertThat(code)
+            .hasLintViolation(5, 2, "Unexpected newline after '.'")
+            .isFormattedAs(formattedCode)
+    }
+
+    @Test
+    fun `Given nested chains which are correctly wrapped`() {
```

Comment:
should we also add TC with index operator like below

```
list(1)[
    0
]
.inc()

list(1)[
    0
].
inc()

list(1)[
    0
]!!
.inc()
```

Reply:
Hmm, this is not handled well (yet). It is formatted as:
```
    list(1)[
        0,
    ]
        .inc()

    list(1)[
        0,
    ]
        ?.inc() // Assumption is that here the `?.` chain operator was intended

    list(1)[
        0,
    ]!!
        .inc()
```

So a TC and fix is needed. I would say that expected outcome should be:
```
    list(1)[
        0,
    ].inc()

    list(1)[
        0,
    ]?.inc() // Assumption is that here the `?.` chain operator was intended

    list(1)[
        0,
    ]!!
        .inc()
```